### PR TITLE
rgw: fix CopyObj crash after admin override

### PR DESCRIPTION
**Pull request by cbodley**
if `RGWCopyObj::verify_permissions()` returns an error, it may leave some zipper objects uninitialized. when the user has admin or system privileges, we'll ignore that error and call `execute()` anyway. this moves the initialization into `RGWCopyObj::init_processing()` instead

Fixes: https://tracker.ceph.com/issues/58035

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
the first commit "dont duplicate sal handles in RGWCopyObj" was enough to resolve the crash on null `dest_bucket`, because `s->bucket` is already initialized

**Comment by cbodley** [comment]
@mattbenjamin i requested your review for the rgw_file change; qa/workunits/rgw/test_librgw_file.sh still passes so i *think* it's correct

File: src/rgw/rgw_file.h [hunk 1]
```diff
@@ -2610,16 +2610,17 @@ class RGWCopyObjRequest : public RGWLibRequest,
...
 
-    src_bucket_name = src_parent->bucket_name();
-    state->src_bucket_name = src_bucket_name;
-    dest_bucket_name = dst_parent->bucket_name();
-    state->bucket_name = dest_bucket_name;
-    dest_obj_name = dst_parent->format_child_name(dst_name, false);
+    state->src_bucket_name = src_parent->bucket_name();
+    state->bucket_name = dst_parent->bucket_name();
+
+    std::string dest_obj_name = dst_parent->format_child_name(dst_name, false);
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.h:2616]: ok

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -5345,45 +5345,30 @@ int RGWCopyObj::verify_permission(optional_yield y)
 
   RGWAccessControlPolicy dest_bucket_policy(s->cct);
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:5356]: assuming this is correct, it's certainly easier to understand

File: src/rgw/rgw_op.cc [hunk 3]
```diff
@@ -488,10 +488,9 @@ int rgw_build_bucket_policies(const DoutPrefixProvider *dpp, rgw::sal::Driver* d
   if (!s->src_bucket_name.empty()) {
     std::unique_ptr<rgw::sal::Bucket> src_bucket;
     ret = driver->get_bucket(dpp, nullptr,
-			    rgw_bucket(s->src_tenant_name,
-				       s->src_bucket_name,
-				       s->bucket_instance_id),
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc:493]: what effect did this mix-up have?

**Review by mattbenjamin — APPROVED** [review]
lgtm

File: src/rgw/rgw_op.cc (hunk 3 above)
Comment (cbodley) [reply src/rgw/rgw_op.cc:493]: it probably never mattered. we only set `bucket_instance_id` if the request included an "rgwx-bucket-instance" header, and we removed multisite's last use of that with the resharding changes. it had only been used for [bucket listings](https://github.com/ceph/ceph/blob/quincy/src/rgw/rgw_data_sync.cc#L3270), not copies

**Review by dang — APPROVED** [review]

**Comment by cbodley** [comment]
update: seeing this tcmalloc crash all over the place in focal testing. vstart fails to bootstrap:
> /home/cbodley/ceph/build/bin/ceph-mon --mkfs -c /home/cbodley/ceph/build/ceph.conf -i a --monmap=/tmp/ceph_monmap.3740 --keyring=/home/cbodley/ceph/build/keyring
src/tcmalloc.cc:332] Attempt to free invalid pointer 0x604000000010
../src/vstart.sh: line 28:  3816 Aborted                 (core dumped) PATH=$CEPH_BIN:$PATH "$@"
from gdb:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50                                                                   
#1  0x00007fffd0e1a859 in __GI_abort () at abort.c:79                                                                                       
#2  0x00007fffd21391d2 in ?? () from /lib/x86_64-linux-gnu/libtcmalloc.so.4                                                                 
#3  0x00007fffd213aca9 in ?? () from /lib/x86_64-linux-gnu/libtcmalloc.so.4                                       
#4  0x00007fffd214fe1d in MallocExtension::Initialize() () from /lib/x86_64-linux-gnu/libtcmalloc.so.4                                      
#5  0x00007fffd2139b1e in ?? () from /lib/x86_64-linux-gnu/libtcmalloc.so.4                                                                 
#6  0x00007ffff7fe0b9a in call_init (l=<optimized out>, argc=argc@entry=8, argv=argv@entry=0x7fffffffdff8, env=env@entry=0x7fffffffe040)    
    at dl-init.c:72                                                                                                                         
#7  0x00007ffff7fe0ca1 in call_init (env=0x7fffffffe040, argv=0x7fffffffdff8, argc=8, l=<optimized out>) at dl-init.c:30                    
#8  _dl_init (main_map=0x7ffff7ffe190, argc=8, argv=0x7fffffffdff8, env=0x7fffffffe040) at dl-init.c:119                                    
#9  0x00007ffff7fd013a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2                                                                
#10 0x0000000000000008 in ?? ()
```
i see the same startup crash from `ceph-osd` and `radosgw` there

https://tracker.ceph.com/issues/58219 tracks a similar crash from the fs suite, but it went away after [reverting](https://github.com/ceph/ceph/pull/49842) a commit that changed ceph-dencoder linkage

**Comment by cbodley** [comment]
@mattbenjamin i'm seeing those same ubuntu20 crashes on main in https://pulpito.ceph.com/cbodley-2023-03-30_18:07:30-rgw:verify-main-distro-default-smithi/

**Comment by mattbenjamin** [comment]
> @mattbenjamin i'm seeing those same ubuntu20 crashes on main in https://pulpito.ceph.com/cbodley-2023-03-30_18:07:30-rgw:verify-main-distro-default-smithi/

that suggests this run was a pass, then?

Matt

**Comment by cbodley** [comment]
> @mattbenjamin i'm seeing those same ubuntu20 crashes on main in https://pulpito.ceph.com/cbodley-2023-03-30_18:07:30-rgw:verify-main-distro-default-smithi/

there were some other failures in the java and datacache tests in https://pulpito.ceph.com/cbodley-2023-03-29_20:47:34-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by cbodley** [comment]
> there were some other failures in the java and datacache tests in https://pulpito.ceph.com/cbodley-2023-03-29_20:47:34-rgw-wip-cbodley-testing-distro-default-smithi/

of those 17 failures, they all look unrelated. i did reruns for each distro:
ubuntu 20.04: https://pulpito.ceph.com/cbodley-2023-03-30_13:17:22-rgw:verify-wip-cbodley-testing-distro-default-smithi/
centos 8: https://pulpito.ceph.com/cbodley-2023-03-30_20:32:49-rgw:verify-wip-cbodley-testing-distro-default-smithi/
ubuntu 22.04: https://pulpito.ceph.com/cbodley-2023-03-30_21:31:09-rgw:verify-wip-cbodley-testing-distro-default-smithi/

only the ubuntu 20.04 failures prevented other tests from running. the centos tests were green, and the ubuntu 22.04 runs contained `POOL_APP_NOT_ENABLED` cluster warnings that will probably be whitelisted in https://github.com/ceph/ceph/pull/47560

i've opened https://tracker.ceph.com/issues/59269 about the tcmalloc crash in librgw_file tests and mailed the dev list. @mkogan1 and i are discussing the datacache failure in https://github.com/ceph/ceph/pull/49328

**Pull request merged**